src/app/archive/archive_lib/processor.ml

```diff
@@ -75,6 +133,78 @@ let rec deferred_result_list_fold ls ~init ~f =
       let%bind init = f init h in
       deferred_result_list_fold t ~init ~f
 
+let deferred_result_list_mapi ~f xs =
+  let open Deferred.Result.Let_syntax in
+  deferred_result_list_fold xs ~init:(0, []) ~f:(fun (index, acc) x ->
+      let%map res = f index x in
+      (Int.succ index, res :: acc))
+  >>| snd >>| List.rev
+
+let deferred_result_list_map ~f = deferred_result_list_mapi ~f:(Fn.const f)
+
+let deferred_result_lift_opt :
+    ('a, 'err) Deferred.Result.t option -> ('a option, 'err) Deferred.Result.t =
+  let open Deferred.Result.Let_syntax in
+  function Some x -> x >>| Option.some | None -> return None
+
+let add_if_some (f : 'arg -> ('res, 'err) Deferred.Result.t) :
+    'arg option -> ('res option, 'err) Deferred.Result.t =
+  Fn.compose deferred_result_lift_opt @@ Option.map ~f
+
+let add_if_set (f : 'arg -> ('res, 'err) Deferred.Result.t) :
+    'arg Snapp_basic.Set_or_keep.t -> ('res option, 'err) Deferred.Result.t =
+  Fn.compose (add_if_some f) Snapp_basic.Set_or_keep.to_option
+
+let add_if_check (f : 'arg -> ('res, 'err) Deferred.Result.t) :
+    'arg Snapp_basic.Or_ignore.t -> ('res option, 'err) Deferred.Result.t =
+  Fn.compose (add_if_some f) Snapp_basic.Or_ignore.to_option
+
+(* `select_cols ~select:"s0" ~table_name:"t0" ["col0", "col1", ...]` creates the string
+   `"SELECT s0 FROM t0 WHERE col0 = ? AND col1 = ? AND..."`.
+   The optional `tannot` function maps column names to type annotations.
+   No type annotation is included if `tannot` returns an empty string. *)
+let select_cols ~(select : string) ~(table_name : string)
```

Comment:
would it be better to use an option type for `tannot`, instead of relying on the empty string to indicate there's no annotation?